lib/mapObjects/CommonConstructors.cpp

```diff
@@ -291,6 +291,69 @@ void BoatInstanceConstructor::configureObject(CGObjectInstance * object, CRandom
 
 }
 
+void MarketInstanceConstructor::initTypeData(const JsonNode & input)
+{
+	for(auto & element : input["modes"].Vector())
+	{
+		if(MappedKeys::MARKET_NAMES_TO_TYPES.count(element.String()))
+			marketModes.insert(MappedKeys::MARKET_NAMES_TO_TYPES.at(element.String()));
+	}
+	
+	marketEfficacy = input["efficacy"].isNull() ? -1 : input["efficacy"].Integer();
+	predefinedOffer = input["offer"];
+	
+	title = input["title"].String();
+	speech = input["speech"].String();
+}
+
+CGObjectInstance * MarketInstanceConstructor::create(std::shared_ptr<const ObjectTemplate> tmpl) const
+{
+	CGMarket * market = nullptr;
+	if(marketModes.size() == 1)
+	{
+		switch(*marketModes.begin())
+		{
+			case EMarketMode::ARTIFACT_RESOURCE:
+			case EMarketMode::RESOURCE_ARTIFACT:
+				market = new CGBlackMarket;
+				break;
+				
+			case EMarketMode::RESOURCE_SKILL:
+				market = new CGUniversity;
+				break;
+		}
+	}
+	
+	if(!market)
+		market = new CGMarket;
+	
+	preInitObject(market);
+
+	if(tmpl)
+		market->appearance = tmpl;
+	market->marketModes = marketModes;
+	if(marketEfficacy >= 0)
+		market->marketEfficacy = marketEfficacy;
+	
+	market->title = market->getObjectName();
+	if(!title.empty())
+		market->title = VLC->generaltexth->translate(title);
```

Comment:
Same here - make it obligatory?

Reply:
There is existing logic that if field is missing, object name will be used